Add new test case for possible bug in BufReader

Milack27 opened the pull request:
When reading a large chunk from a BufReader, if all the bytes from the buffer have been already consumed, the internal buffer is bypassed entirely. However, it is not invalidated, and it's possible to access its contents using the `seek_relative` method, because it tries to reuse the existing buffer.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @JoshTriplett (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/182074750) ([raw log](https://api.travis-ci.com/v3/job/182074750/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:13900518:start=1551707147617404345,finish=1551707148471869957,duration=854465612
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
[00:03:49] tidy error: /checkout/src/libstd/io/buffered.rs:1071: trailing whitespace
[00:03:50] some tidy checks failed
[00:03:50] 
[00:03:50] 
[00:03:50] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:50] 
[00:03:50] 
[00:03:50] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:50] Build completed unsuccessfully in 0:00:46
[00:03:50] Build completed unsuccessfully in 0:00:46
[00:03:51] make: *** [tidy] Error 1
[00:03:51] Makefile:68: recipe for target 'tidy' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0585b86e
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Mar  4 13:49:49 UTC 2019
---
travis_time:end:01791600:start=1551707390612094975,finish=1551707390617267310,duration=5172335
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:16fac448
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:125de45b
travis_time:start:125de45b
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:012e6994
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/182081588) ([raw log](https://api.travis-ci.com/v3/job/182081588/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:155f79e0:start=1551708531382670248,finish=1551708532305729200,duration=923058952
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:28] 
[01:21:28] running 119 tests
[01:21:55] .iiiii...i.....i..i...i..i.i..i.ii...i.....i..i....i..........iiii..........i...ii...i.......ii.i.i. 100/119
[01:21:59] i......iii.i.....ii
[01:21:59] 
[01:21:59]  finished in 31.395
[01:21:59] travis_fold:end:test_debuginfo

---
[01:38:58] travis_fold:start:test_stage1-std
travis_time:start:test_stage1-std
Testing std stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:38:59]    Compiling std v0.0.0 (/checkout/src/libstd)
[01:39:01] error[E0433]: failed to resolve: use of undeclared type or module `RevBufReader`
[01:39:01]      |
[01:39:01]      |
[01:39:01] 1075 |         let mut reader = RevBufReader::with_capacity(3, io::Cursor::new(inner));
[01:39:01]      |                          ^^^^^^^^^^^^ use of undeclared type or module `RevBufReader`
[01:39:13] error: aborting due to previous error
[01:39:13] 
[01:39:13] For more information about this error, try `rustc --explain E0433`.
[01:39:13] error: Could not compile `std`.
[01:39:13] error: Could not compile `std`.
[01:39:13] 
[01:39:13] To learn more, run the command again with --verbose.
[01:39:13] 
[01:39:13] 
[01:39:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "std" "--" "--quiet"
[01:39:13] 
[01:39:13] 
[01:39:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:39:13] Build completed unsuccessfully in 0:29:57
[01:39:13] Build completed unsuccessfully in 0:29:57
[01:39:13] Makefile:48: recipe for target 'check' failed
[01:39:13] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0ced8d38
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Mar  4 15:48:16 UTC 2019
Mon Mar  4 15:48:16 UTC 2019
echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:14fcb930
travis_time:start:14fcb930
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0d08b418
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/182120802) ([raw log](https://api.travis-ci.com/v3/job/182120802/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0c9b8eaf:start=1551715561873205141,finish=1551715563775151375,duration=1901946234
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:20:17] 
[01:20:17] running 119 tests
[01:20:46] .iiiii...i.....i..i...i..i.i..i.ii...i.....i..i....i..........iiii..........i...ii...i.......ii.i.i. 100/119
[01:20:51] i......iii.i.....ii
[01:20:51] 
[01:20:51]  finished in 33.384
[01:20:51] travis_fold:end:test_debuginfo

---
[01:39:01] .................................................................................................... 100/774
[01:39:01] .................................................................................................... 200/774
[01:39:01] thread '<unnamed>' panicked at 'explicit panic', src/libstd/io/buffered.rs:1339:17
[01:39:01] thread '<unnamed>' panicked at 'explicit panic', src/libstd/io/stdio.rs:805:13
[01:39:01] ......................................................F............................................. 300/774
[01:39:02] .................................................................................................... 500/774
[01:39:02] thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:997:5
[01:39:02] thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', src/libcore/result.rs:997:5
[01:39:02] thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', src/libcore/result.rs:997:5
---
[01:39:13] failures:
[01:39:13] 
[01:39:13] ---- io::buffered::tests::test_buffered_reader_seek_relative_read stdout ----
[01:39:13] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:39:13]   left: `[6, 7]`,
[01:39:13]  right: `[3, 4]`', src/libstd/io/buffered.rs:1087:9
[01:39:13] 
[01:39:13] failures:
[01:39:13]     io::buffered::tests::test_buffered_reader_seek_relative_read
[01:39:13] 
[01:39:13] 
[01:39:13] test result: FAILED. 773 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
[01:39:13] 
[01:39:13] error: test failed, to rerun pass '--lib'
[01:39:13] 
[01:39:13] 
[01:39:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "std" "--" "--quiet"
[01:39:13] 
[01:39:13] 
[01:39:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:39:13] Build completed unsuccessfully in 0:31:25
[01:39:13] Build completed unsuccessfully in 0:31:25
[01:39:13] Makefile:48: recipe for target 'check' failed
[01:39:13] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0019d106
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon Mar  4 17:45:28 UTC 2019
---
travis_time:end:0011346a:start=1551721530381457301,finish=1551721530386629441,duration=5172140
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:036a4ca4
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(e
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

joshtriplett commented:
I think the test case makes sense; if the buffer still exists, seeking to a location within it seems reasonable.

I don't think it makes sense to throw away the buffer until it gets overwritten. But once it *does* get overwritten, it needs invalidating.

Do you have a test case in which you can read data from the buffer that you shouldn't be able to?

Milack27 commented:
@joshtriplett Yes, I believe the test case that I presented already shows that it is possible to access data from the buffer when we shouldn't:

1. `[5, 6, 7]` is read and stored in the buffer;
1. `[0, 1, 2, 3, 4]` is read bypassing the buffer. The buffer should now be invalidated but it's not, the chunk `[5, 6, 7]` remains there. At this point, we're at the end of the stream.
1. We call seek(-2). The buffer is 3 bytes long, so BufReader understands that we're still in its range, and that it doesn't need to fetch data from the inner reader.
1. We read again, but instead of getting `[3, 4]` the test fails and we get `[6, 7]`, which was incorrectly stored in the buffer.

Milack27 commented:
By the way, sorry for the mess. I'd like to squash those mistaken commits and force push them to keep it organized, but I don't have good internet access right now. I didn't even clone the Rust repository, I edited the file and commited on GitHub itself.

joshtriplett commented:
On Mon, Mar 04, 2019 at 03:14:57PM -0800, André Vicente Milack wrote:
> @joshtriplett Yes, I believe the test case that I presented already shows that it is possible to access data from the buffer when we shouldn't:
> 
> 1. `[5, 6, 7]` is read and stored in the buffer;
> 1. `[0, 1, 2, 3, 4]` is read bypassing the buffer. The buffer should now be invalidated but it's not, the chunk `[5, 6, 7]` remains there. At this point, we're at the end of the stream.
> 1. We call seek(-2). The buffer is 3 bytes long, so BufReader understands that we're still in its range, and that it doesn't need to fetch data from the inner reader.
> 1. We read again, but instead of getting `[3, 4]` the test fails and we get `[6, 7]`, which was incorrectly stored in the buffer.

Ah, I see! And does the test pass if you uncomment the commented lines?

If so, please do go ahead and uncomment it and document the reason for
it, and I'd be happy to review that.

Milack27 commented:
Yes, the test passes when we uncomment those lines. I'll do it as soon as possible.

Milack27 commented:
@joshtriplett I realized that this problem not only happened in the `read` method, but also in the `seek` method. I just fixed the problem on both methods (I hope I didn't forget anything this time) and added another test case to cover `seek` as well.

joshtriplett commented:
I feel like I need an extra pair of eyes here. I can see how this makes sense, but it feels like it might be invalidating more than it needs to, and I also wonder why it looks like seek's control flow can pass through both the invalidates you added in one call.

@Diggsey, @sfackler, it looks like you've done a fair bit of work in this area, can you take a look, please?

Milack27 commented:
> [...] and I also wonder why it looks like seek's control flow can pass through both the invalidates you added in one call.

I had the same question when I first read that code. It seems to me that the invalidation inside the else is there because the seek just below it can fail and return without passing through the invalidation outside the else.

Diggsey commented on src/libstd/io/buffered.rs:
```diff
@@ -323,14 +326,18 @@ impl<R: Seek> Seek for BufReader<R> {
-                self.pos = self.cap; // empty the buffer
+                // Empty the buffer
+                self.cap = 0;
+                self.pos = 0;
```
Why do we need to mark the buffer as empty here, when we do the same further down?

Diggsey replied:
Never-mind just saw the conversation above!

Diggsey commented:
What about `read_vectored` - I notice that doesn't even attempt to clear the buffer?

Diggsey commented:
Also, it might be worth pulling this out into a method, `discard_buffer()` so that this mistake isn't repeated.

Milack27 commented:
@Diggsey Good point, I didn't understand yet how `read_vectored` works, so I didn't touch it. And I like the `discard_buffer()` idea, I'll do it.

Diggsey commented:
It's pretty much the same as `read()`, just using multiple user-provided buffers instead of one. It needs a call to `discard_buffer()` in the same place as for the `read()` call.

Milack27 commented:
Does `discard_buffer` need any other annotation?

```Rust
   /// Invalidates all data in the internal buffer.
    #[inline]
    fn discard_buffer(&mut self) {
        self.pos = 0;
        self.cap = 0;
    }
```

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/182849429) ([raw log](https://api.travis-ci.com/v3/job/182849429/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0bf9a846:start=1551900765085032461,finish=1551900766017839675,duration=932807214
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
Setting environment variables from .travis.yml
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Milack27 commented:
I had tried to improve the readability of the seek method, but I ended up changing its behavior as well, so I just rolled back.

Dylan-DPC commented:
ping from triage @joshtriplett any updates on this?

joshtriplett commented:
@bors r+

bors commented:
:pushpin: Commit c36d91c5cc446a4688186be1071e0e139ba9d38f has been approved by `joshtriplett`

<!-- @bors r=joshtriplett c36d91c5cc446a4688186be1071e0e139ba9d38f -->

bors commented:
:hourglass: Testing commit c36d91c5cc446a4688186be1071e0e139ba9d38f with merge 15a5dfa0b4438d199e8703fd578f59a5049527be...
<!-- homu: {"type":"BuildStarted","head_sha":"c36d91c5cc446a4688186be1071e0e139ba9d38f","merge_sha":"15a5dfa0b4438d199e8703fd578f59a5049527be"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/105236505), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/23243155)
Approved by: joshtriplett
Pushing 15a5dfa0b4438d199e8703fd578f59a5049527be to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joshtriplett","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/23243155","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/105236505"},"merge_sha":"15a5dfa0b4438d199e8703fd578f59a5049527be"} -->
